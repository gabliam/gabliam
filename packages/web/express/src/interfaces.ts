import { DefaultMethods } from '@gabliam/web-core';
import { express } from './express';

/**
 * Represent a method that create an express router
 */
export type RouterCreator = () => express.Router;

export type ExpressMethods = Extract<keyof express.Router, DefaultMethods>;
