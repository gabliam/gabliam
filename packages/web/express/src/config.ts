import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import bodyParser from 'body-parser';
import { EXPRESS_PLUGIN_CONFIG } from './constants';

const options = Joi.object().keys({
  inflate: Joi.boolean(),
  limit: [Joi.number(), Joi.string()],
  type: [Joi.string(), Joi.array().items(Joi.string())],
});

const optionsJson = options.keys({
  strict: Joi.boolean(),
});

const optionsText = options.keys({
  defaultCharset: Joi.string(),
});

const optionsUrlencoded = options
  .keys({
    extended: Joi.boolean(),
    parameterLimit: Joi.number(),
  })
  .default({ extended: false });

const configJoi = Joi.object()
  .keys({
    json: Joi.boolean().default(true),
    optionsJson,
    urlencoded: Joi.boolean().default(true),
    optionsUrlencoded,
    text: Joi.boolean().default(false),
    optionsText,
  })
  .default({
    json: true,
    urlencoded: true,
    optionsUrlencoded: { extended: false },
    text: false,
  });

export interface BodyParserConfig {
  json: boolean;
  optionsJson?: bodyParser.OptionsJson;
  urlencoded: boolean;
  optionsUrlencoded?: bodyParser.OptionsUrlencoded;
  text: boolean;
  optionsText?: bodyParser.OptionsText;
}

export interface ExpressConfig {
  bodyParser: BodyParserConfig;
}

@PluginConfig()
export class ExpressPluginConfig implements ExpressConfig {
  @Value('application.web.bodyParser', configJoi)
  bodyParser: BodyParserConfig;

  @Bean(EXPRESS_PLUGIN_CONFIG)
  restConfig(): ExpressConfig {
    return {
      bodyParser: this.bodyParser,
    };
  }
}
