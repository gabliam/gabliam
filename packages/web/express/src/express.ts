import express from 'express';
import { GabContext } from '@gabliam/web-core';

export { express };

export interface CustomRequest extends express.Request {
  context: GabContext;
}
