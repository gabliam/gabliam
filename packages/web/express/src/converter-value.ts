import {
  ExecutionContext,
  GabContext,
  ResponseEntity,
  convertValueFn,
} from '@gabliam/web-core';
import { express } from './express';

export const converterValue: convertValueFn = (
  ctx: GabContext,
  execCtx: ExecutionContext,
  result: any,
) => {
  // response handler if the result is a ResponseEntity
  function convertEntity(value: ResponseEntity) {
    if (value.hasHeader()) {
      ctx.response.set(<any>value.headers);
    }
    ctx.status = value.status;
    ctx.body = value.body;
  }

  if (!ctx.headersSent) {
    if (result !== undefined) {
      if (result instanceof ResponseEntity) {
        convertEntity(result);
      } else if (typeof result === 'string' || typeof result === 'object') {
        ctx.body = result;
      } else {
        ctx.body = `${result}`;
      }
    }
  }
};

export const send = (ctx: GabContext, res: express.Response, json: boolean) => {
  const { status, message, body, type } = ctx;
  /* istanbul ignore next */
  if (type) {
    res.type(type);
  }
  /* istanbul ignore next */
  if (message) {
    res.statusMessage = message;
  }

  if (status) {
    res.status(status);
  }
  if (json) {
    res.json(body);
  } else {
    res.send(body);
  }
};
