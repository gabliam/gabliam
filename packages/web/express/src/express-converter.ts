import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  BadInterceptorError,
  createInterceptorResolver,
  getContext,
  getExtractArgs,
  isInterceptor,
} from '@gabliam/web-core';
import { express } from './express';

/**
 * Class for convert interceptor to Express Middleware
 */
@InjectContainer()
@Service()
export class ExpressConverter {
  public interceptorToMiddleware(clazz: any): express.RequestHandler {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new BadInterceptorError(instance);
    }

    const extractArgs = getExtractArgs(container, instance, 'intercept', true);

    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction,
    ) => {
      const args = await extractArgs(getContext(req), null, next);

      try {
        await toPromise(instance.intercept(...args));
        next();
      } catch (err) {
        next(err);
      }
    };
  }
}
