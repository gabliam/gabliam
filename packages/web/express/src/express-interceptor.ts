import { Service } from '@gabliam/core';
import {
  Context,
  GabContext,
  Interceptor,
  InterceptorConstructor,
  METADATA_KEY,
  Next,
  nextFn,
  UseInterceptors,
} from '@gabliam/web-core';
import { express } from './express';

export type ExpressInterceptorType = 'RequestHandler' | 'intercept';

/**
 * Test if target is an express interceptor
 * @param target any
 */
export const isValidInterceptor = (target: any) => {
  const meta = Reflect.getOwnMetadata(
    METADATA_KEY.specialInterceptor,
    target.constructor || target,
  );

  return meta === undefined || meta === 'express';
};

/**
 * ExpressInterceptor decorator
 * class is an express interceptor.
 * Must return express middleware
 */
const ExpressInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.specialInterceptor, 'express', target);
};

/**
 * Convert a Express middleware to an express interceptor
 */
export const toInterceptor = (
  mid: express.RequestHandler,
): InterceptorConstructor => {
  const clazz: InterceptorConstructor = class implements Interceptor {
    async intercept(context: GabContext, next: nextFn): Promise<any> {
      const req = context.request.originalRequest;
      const res = context.response.originalResponse;
      switch (mid.length) {
        // with callback
        case 3:
          return new Promise((resolve, reject) => {
            (<express.RequestHandler>mid)(req, res, (err: any) => {
              if (err) {
                reject(err);
              } else {
                resolve(next());
              }
            });
          });
        // with error
        case 2:
        default:
          (<any>mid)(req, res);
          await next();
          return Promise.resolve();
      }
    }
  };
  Context()(clazz.prototype, 'intercept', 0);
  Next()(clazz.prototype, 'intercept', 1);
  Service()(clazz);
  ExpressInterceptor()(clazz);

  // metadata
  Reflect.defineMetadata('design:type', Function, clazz.prototype, 'intercept');
  Reflect.defineMetadata(
    'design:paramtypes',
    [Function, Function],
    clazz.prototype,
    'intercept',
  );
  Reflect.defineMetadata(
    'design:returntype',
    Promise,
    clazz.prototype,
    'intercept',
  );
  return clazz;
};

/**
 * Alias for evict to use: UseInterceptors(toInterceptor(expressMid))
 * /!\ Use it for RequestHandler
 */
export const UseExpressInterceptors = (...mids: express.RequestHandler[]) =>
  UseInterceptors(...mids.map((m) => toInterceptor(m)));
