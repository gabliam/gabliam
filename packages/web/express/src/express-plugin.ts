import {
  Container,
  GabliamPlugin,
  Plugin,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import {
  APP,
  cleanPath,
  compose,
  ExecutionContext,
  getContext,
  InterceptorInfo,
  REQUEST_LISTENER_CREATOR,
  RestMetadata,
  SERVER,
  WebConfiguration,
  WebConfigurationContructor,
  WebPluginBase,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
} from '@gabliam/web-core';
import d from 'debug';
import http from 'http';
import { CUSTOM_ROUTER_CREATOR } from './constants';
import { converterValue, send } from './converter-value';
import { express } from './express';
import { isValidInterceptor } from './express-interceptor';
import { ExpressMethods, RouterCreator } from './interfaces';
import { addContextMiddleware, addMiddlewares } from './middleware';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('ExpressPlugin')
@Scan()
export class ExpressPlugin extends WebPluginBase<express.Application>
  implements GabliamPlugin {
  constructor(
    config?: Partial<WebConfigurationContructor<express.Application>>
  ) {
    super(config);
  }

  bindApp(
    container: Container,
    registry: Registry,
    webConfiguration: WebConfiguration
  ): void {
    container.bind(APP).toConstantValue(express());
    container.bind(REQUEST_LISTENER_CREATOR).toConstantValue(() => {
      const app = container.get<express.Application>(APP);
      return app;
    });
    webConfiguration.addwebConfig({
      instance: addMiddlewares,
      order: -2,
    });

    webConfiguration.addwebConfig({
      instance: addContextMiddleware,
      order: -1,
    });
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    app.set('port', restConfig.port);
    await super.start(container, registry);
  }

  async buildControllers(
    restMetadata: RestMetadata<ExpressMethods>,
    container: Container
  ) {
    const app = container.get<express.Application>(APP);

    // get the router creator
    let routerCreator: RouterCreator = () => express.Router();
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    for (const [
      controllerId,
      { methods, controllerPath },
    ] of restMetadata.controllerInfo) {
      const controller = container.get<object>(controllerId);
      const router = routerCreator();
      const routerPath = cleanPath(`${restMetadata.rootPath}${controllerPath}`);

      debug(`New route : "${routerPath}"`);

      for (const methodInfo of methods) {
        const execCtx = new ExecutionContext(controller, methodInfo);

        const interceptors = methodInfo.interceptors.filter(i =>
          isValidInterceptor(i)
        );

        // create handler
        const handler = this.handlerFactory(execCtx, interceptors);

        // register handler in router
        router[methodInfo.method](methodInfo.methodPath, handler);
      }
      app.use(routerPath, router);
    }
  }

  private handlerFactory(
    execCtx: ExecutionContext,
    interceptors: InterceptorInfo[]
  ): express.RequestHandler {
    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const composeInterceptor = compose(
        interceptors,
        converterValue
      );
      const ctx = getContext(req);
      const methodInfo = execCtx.getMethodInfo();
      const controller = execCtx.getClass();

      let expressNextCall = false;
      const expressNext = () => {
        expressNextCall = true;
        next();
      };

      const callNext = async () => {
        const args = await toPromise(
          methodInfo.extractArgs(ctx, execCtx, expressNext)
        );
        return await toPromise(controller[methodInfo.methodName](...args));
      };

      try {
        await composeInterceptor(ctx, execCtx, callNext);
        if (!expressNextCall) {
          send(ctx, res, methodInfo.json);
        }
      } catch (err) {
        next(err);
      }
    };
  }
}
