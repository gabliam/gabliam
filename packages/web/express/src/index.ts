import { ExpressPlugin } from './express-plugin';

export * from 'http-status-codes';
export * from './express';
export * from './express-converter';
export * from './express-interceptor';
export * from './interfaces';
export * from './request';
export * from './response';
export { ExpressPlugin as default };
