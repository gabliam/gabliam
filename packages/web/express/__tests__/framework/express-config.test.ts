import { Config } from '@gabliam/core';
import {
  Controller,
  Get,
  WebConfig,
  WebConfigAfterControllers,
} from '@gabliam/web-core';
import { WebPluginTest } from '@gabliam/web-core/src/testing';
import sinon from 'sinon';
import ExpressConfig, { express as e } from '../../src';

let appTest: WebPluginTest;
let result: string;
const middleware: any = {
  a(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
    result += 'a';
    nextFunc();
  },
  b(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
    result += 'b';
    nextFunc();
  },
  c(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
    result += 'c';
    nextFunc();
  },
  e(error: any, req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
    result += 'e';
    res.status(500).send('error');
  },
  d(error: any, req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
    result += 'd';
    nextFunc(error);
  },
};
const spyA = sinon.spy(middleware, 'a');
const spyB = sinon.spy(middleware, 'b');
const spyC = sinon.spy(middleware, 'c');
const spyE = sinon.spy(middleware, 'e');
const spyD = sinon.spy(middleware, 'd');

beforeEach(async () => {
  result = '';
  spyA.resetHistory();
  spyB.resetHistory();
  spyC.resetHistory();
  spyE.resetHistory();
  spyD.resetHistory();
  appTest = new WebPluginTest([ExpressConfig]);
});

afterEach(async () => {
  await appTest.destroy();
});

test('@ExpressConfig', async () => {
  @Controller('/')
  class TestController {
    @Get('/')
    public getTest() {
      return 'GET';
    }
  }

  @Config()
  class ServerConfig {
    @WebConfig()
    serverConfig(app: e.Application) {
      app.use(spyA);
      app.use(spyB);
      app.use(spyC);
    }
  }

  appTest.addClass(TestController);
  appTest.addClass(ServerConfig);
  await appTest.buildAndStart();

  const response = await appTest.supertest().get('/').expect(200);
  expect(spyA.calledOnce).toBe(true);
  expect(spyB.calledOnce).toBe(true);
  expect(spyC.calledOnce).toBe(true);
  expect(response).toMatchSnapshot();
  expect(result).toMatchSnapshot();
});

test('@ExpressConfig Order', async () => {
  @Controller('/')
  class TestController {
    @Get('/')
    public getTest() {
      return 'GET';
    }
  }

  @Config()
  class ServerConfig {
    @WebConfig(150)
    serverConfigC(app: e.Application) {
      app.use(spyC);
    }

    @WebConfig(100)
    serverConfigB(app: e.Application) {
      app.use(spyB);
    }

    @WebConfig(50)
    serverConfigA(app: e.Application) {
      app.use(spyA);
    }
  }

  appTest.addClass(TestController);
  appTest.addClass(ServerConfig);
  await appTest.buildAndStart();

  const response = await appTest.supertest().get('/').expect(200);
  expect(spyA.calledOnce).toBe(true);
  expect(spyB.calledOnce).toBe(true);
  expect(spyC.calledOnce).toBe(true);
  expect(response).toMatchSnapshot();
  expect(result).toMatchSnapshot();
});

test('ErrorConfig', async () => {
  @Controller('/')
  class TestController {
    @Get('/')
    public getTest() {
      throw new Error();
    }
  }

  @Config()
  class ServerConfig {
    @WebConfigAfterControllers()
    serverConfig(app: e.Application) {
      app.use(spyE);
    }
  }

  appTest.addClass(TestController);
  appTest.addClass(ServerConfig);
  await appTest.buildAndStart();

  const response = await appTest.supertest().get('/').expect(500);
  expect(spyE.calledOnce).toBe(true);
  expect(response).toMatchSnapshot();
  expect(result).toMatchSnapshot();
});

test('ErrorConfig order', async () => {
  @Controller('/')
  class TestController {
    @Get('/')
    public getTest() {
      throw new Error();
    }
  }

  @Config()
  class ServerConfig {
    @WebConfigAfterControllers(100)
    serverConfig(app: e.Application) {
      app.use(spyE);
    }

    @WebConfigAfterControllers(50)
    serverConfigD(app: e.Application) {
      app.use(spyD);
    }
  }

  appTest.addClass(TestController);
  appTest.addClass(ServerConfig);
  await appTest.buildAndStart();

  const response = await appTest.supertest().get('/').expect(500);
  expect(spyE.calledOnce).toBe(true);
  expect(spyD.calledOnce).toBe(true);
  expect(response).toMatchSnapshot();
  expect(result).toMatchSnapshot();
});
