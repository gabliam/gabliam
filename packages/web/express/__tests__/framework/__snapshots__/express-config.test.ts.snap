// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ExpressConfig 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`@ExpressConfig 2`] = `"abc"`;

exports[`@ExpressConfig Order 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`@ExpressConfig Order 2`] = `"abc"`;

exports[`ErrorConfig 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 500,
  "text": "error",
}
`;

exports[`ErrorConfig 2`] = `"e"`;

exports[`ErrorConfig order 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 500,
  "text": "error",
}
`;

exports[`ErrorConfig order 2`] = `"de"`;
